be/src/common/configbase.cpp

```diff
@@ -200,202 +137,124 @@ bool Properties::load(const char* filename) {
-bool update(const std::string& value, T& retval) {
-    std::string valstr(value);
-    trim(valstr);
-    if (!replaceenv(valstr)) {
+bool Field::set_value(std::string value) {
+    if (auto st = replaceenv(value); !st.ok()) {
         return false;
     }
-    return strtox(valstr, retval);
+    StripWhiteSpace(&value);
+    return parse_value(value);
 }
 
-template <typename T>
-std::ostream& operator<<(std::ostream& out, const std::vector<T>& v) {
-    size_t last = v.size() - 1;
-    for (size_t i = 0; i < v.size(); ++i) {
-        out << v[i];
-        if (i != last) {
-            out << ", ";
+// Init conf fields.
+bool init(const char* filename) {
+    std::ifstream input;
+    if (filename != nullptr) {
+        input.open(filename);
+        if (input.fail()) {
+            std::cerr << "Fail to open " << filename << std::endl;
+            return false;
         }
     }
-    return out;
+    return init(input);
 }
 
-#define SET_FIELD(FIELD, TYPE)                                                                     \
-    if (strcmp((FIELD).type, #TYPE) == 0) {                                                        \
-        if (!props.get((FIELD).name, (FIELD).defval, *reinterpret_cast<TYPE*>((FIELD).storage))) { \
-            std::cerr << "config field error: " << (FIELD).name << std::endl;                      \
-            return false;                                                                          \
-        }                                                                                          \
-        continue;                                                                                  \
+inline bool init_from_default_values() {
+    for (const auto& [name, field] : Field::fields()) {
+        if (!field->set_value(field->defval())) {
+            std::cerr << fmt::format("Invalid default value of config '{}': '{}'\n", name, field->defval());
+            return false;
+        }
     }
+    return true;
+}
 
-// Init conf fields.
-bool init(const char* filename) {
-    // Load properties file.
-    if (!props.load(filename)) {
+bool init(std::istream& input) {
+    if (!init_from_default_values()) {
         return false;
     }
 
-    // Set conf fields.
-    for (const auto& it : *Register::_s_field_map) {
-        SET_FIELD(it.second, bool);
-        SET_FIELD(it.second, int16_t);
-        SET_FIELD(it.second, int32_t);
-        SET_FIELD(it.second, int64_t);
-        SET_FIELD(it.second, double);
-        SET_FIELD(it.second, std::string);
-        SET_FIELD(it.second, MutableString);
-        SET_FIELD(it.second, std::vector<bool>);
-        SET_FIELD(it.second, std::vector<int16_t>);
-        SET_FIELD(it.second, std::vector<int32_t>);
-        SET_FIELD(it.second, std::vector<int64_t>);
-        SET_FIELD(it.second, std::vector<double>);
-        SET_FIELD(it.second, std::vector<std::string>);
+    std::map<std::string, std::string> configs;
+    if (!parse_key_value_pairs(input, &configs)) {
```

Comment:
should use list instead of set. the order of the parser matters when alias is introduced

e.g.

```
CONF_Alias(cfg_a, alias_aa);
CONF_Alias(cfg_a, alias_zz);
```

in the conf file, it could be
```
alias_zz = 1;
...
alias_aa = 2;
```

`alias_aa` should win at the end. while parsing and save into map, it will be `alias_zz` win due to its alphabetic order.

Reply:
Fixed, PTAL